src/widget/input/completion.rs

```diff
@@ -164,18 +185,81 @@ impl Completion {
         self.selection = Selection::None;
     }
 
-    pub fn process(&mut self, input: &str) {
-        let Some((head, rest)) = input.split_once('/') else {
+    /// Convert a list of User structs into Completion Entries
+    pub fn with_users(&mut self, users: Vec<User>) -> Self {
+        self.users = users
+            .iter()
+            .map(|u| {
+                // Let's make some eternal strings
+                // TODO! Is this bad?
+                let static_str = Box::leak(u.nickname().to_string().into_boxed_str());
+                Entry {
+                    title: static_str,
+                    args: vec![],
+                    completion_type: CompletionType::User,
+                }
+            })
+            .collect();
+        self.clone()
+    }
+
+    /// Determine if the current text is a viable candidate for completion
+    /// If the command starts with a `/`, we can guess that the user is attempting
+    /// to enter a /command.
+    ///
+    /// If the text doesn't have a `/`, then we want to rsplit the words so we can
+    /// get the last word in the input, which can potentially be autocompleted
+    fn valid_completion_input(input: &str, maybe_command: bool) -> Option<(&str, &str)> {
+        if maybe_command {
+            input.split_once('/')
+        } else {
+            // Are we looking at multiple words? Look at the last one
+            match input.rsplit_once(' ') {
+                Some((head, rest)) => Some((head, rest)),
+                // Single command
+                None => Some(("", input)),
+            }
+        }
+    }
+
+    /// If we are autocompleting a word, we want to replace the word we're completing with the
+    /// completion target. This way, we can continue typing if we've, say, completed a users name
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// let input = "Hello my good friend Ano";
+    /// let completion = "AnonymousUser";
+    /// let result = Completion::complete_selected_word(input, completion);
+    /// assert!(result == "Hello my good friend AnonymousUser".to_string());
+    /// ```
+    pub fn complete_selected_word(input: &str, completion: &str) -> String {
+        let Some((original_input, _completion_word)) = input.rsplit_once(' ') else {
+            return completion.into()
+        };
+        format!("{} {}", original_input, completion)
+    }
+
+    pub fn process(&mut self, input: &str, users: Vec<User>) {
+        self.with_users(users);
+        let maybe_command = input.starts_with('/');
+        let Some((head, rest)) = Self::valid_completion_input(input, maybe_command) else {
             self.reset();
             return;
```

Comment:
Let's add some better separation between processing for a command vs user completion. If input starts with a slash, we can branch into a `process_command` function, otherwise `process_user`.

Then we only need to add the users info inside `process_user` compared to us always adding it here